pkg/cmd/pulumi/util.go

```diff
@@ -634,6 +643,63 @@ func getUpdateMetadata(msg, root, execKind, execAgent string, updatePlan bool) (
 	return m, nil
 }
 
+// addPulumiCLIMetadataToEnvironment enriches updates with metadata to provide context about
+// the system, environment, and options that an update is being performed with.
+func addPulumiCLIMetadataToEnvironment(env map[string]string, cmd *cobra.Command, getEnviron func() []string) {
+	// Pulumi Environment Variables (Name and Set/Truthiness Only)
+
+	// Set up a map of all known bool environment variables.
+	boolVars := map[string]struct{}{}
+	for _, e := range declared.Variables() {
+		_, ok := e.Value.(declared.BoolValue)
+		if !ok {
+			continue
+		}
+		boolVars[e.Name()] = struct{}{}
+	}
+
+	// Add all environment variables that start with "PULUMI_" and their set-ness or truthiness.
+	for _, envvar := range getEnviron() {
+		if !strings.HasPrefix(envvar, "PULUMI_") {
+			// Not a Pulumi environment variable. Skip.
+			continue
+		}
+		s := strings.SplitN(envvar, "=", 2)
+		if len(s) < 2 {
+			// Invalid environment variable. Skip.
+			continue
+		}
+		name, value := s[0], s[1]
+		flag := "set"
+		if _, isBoolVar := boolVars[name]; isBoolVar {
+			flag = "false"
+			if cmdutil.IsTruthy(value) {
+				flag = "true"
+			}
+		}
+
+		// Only store the variable name and set-ness or truthiness not the value.
+		env["pulumi.env."+name] = flag
+	}
+
+	// System information
+	env["pulumi.version"] = version.Version
+	env["pulumi.os"] = runtime.GOOS
+	env["pulumi.arch"] = runtime.GOARCH
+
+	// Guard against nil pointer dereference.
+	if cmd == nil {
+		// No command was provided. Don't add flags.
+		logging.V(3).Infof("no command was provided to addPulumiCLIMetadataToEnvironment; not adding flags")
+		return
+	}
+
+	// Pulumi CLI Flags (Name Only)
+	cmd.Flags().Visit(func(f *pflag.Flag) {
+		env["pulumi.flag."+f.Name] = "set"
```

Comment:
Can we do a similar thing for bool flags to check if they're true/false and send that instead of just "set".

Reply:
@dixler FYI, for this, you can use [FlagSet.GetBool](https://pkg.go.dev/github.com/spf13/pflag#FlagSet.GetBool) to get the value of the flag. It'll return an error if the flag is not a boolean flag, in which case you can use "set".